### PR TITLE
[Share 2.0] register providers

### PR DESCRIPTION
**Pull request by rullzer**
For #19331

Allow registering of providers.

Lets see what CI has to say.

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
Will there ever be the need to have multiple providers of the same type?

If not we can move this to pure config as we did this with the commants manager
https://github.com/owncloud/core/blob/master/lib/private/comments/managerfactory.php
https://github.com/owncloud/core/blob/master/config/config.sample.php#L812

Would result in less code - and less to test ....

**Comment by rullzer** [comment]
@DeepDiver1975 no I don't think so. Because if there are multiple providers for the same type how will we know which once to use.

We will need multiple providers. And multiple types per provider.

Moving to config is also fine with me...

@schiesbn opinion?

**Comment by DeepDiver1975** [comment]
I don't think we need mutliple providers for the same share type.

From my POV this is beyond the current concept to have 10 implementations of public link sharing running at the same time.

**Comment by rullzer** [comment]
@DeepDiver1975 yeah I'm agreeing with you :P

**Comment by schiesbn** [comment]
I don't think we need to have multiple providers for the same share type. I would only allow to register one provider for one share type. Even if, at some point in time, we have two share provider which do some kind of link sharing they should have a different share type. For example "linkShare" and "SuperLinkShareNG". The share type should the the unique identifier which connects a share to a share provider.

**Comment by DeepDiver1975** [comment]
so - reading the classname from the config is much easier from my pov

**Comment by rullzer** [comment]
Ok let me look into that then :)

**Comment by rullzer** [comment]
Ok so after looking into it. I think it having this into config.php makes sense. Since it requires explicit action from the admin to setup the different share providers. Which I think is a good thing here.

But since we will have 2 share providers at least (default and federated) I'm thinking how to represent this. And how not to complecate the setup to much. Because we need to know for which share types.

So in short the code needs to stay I think. But it reduces stuff in the public name space.

**Comment by DeepDiver1975** [comment]
> But since we will have 2 share providers at least (default and federated) I'm thinking how to represent this. And how not to complecate the setup to much. Because we need to know for which share types.

I'd leave this to the factory - the factory is the class which is defined in the config.

The factory method will return the provider for the asked share type.

**Comment by rullzer** [comment]
@DeepDiver1975 you mean that we list in general 2 provider factories there.
Then we call those factories and they return the class and an array of share types which wee will then register at the manager?

I can see that working :)

**Comment by DeepDiver1975** [comment]
One factory to rule them all.

**Comment by rullzer** [comment]
> One factory to rule them all. 

But if people want to deploy their own sharae provider it is likely they only want to do the 'main' share provider. But keep the federated sharing for example. Would they then need to create their own factory to create their own provider and the federated provider?

**Comment by rullzer** [comment]
Now using the factory apporoach. Everything still works as expected.

File: lib/private/share20/iproviderfactory.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OC\Share20;
```
Comment (nickvergessen) [inline lib/private/share20/iproviderfactory.php]: should  this be public?

Comment (rullzer) [inline lib/private/share20/iproviderfactory.php]: yes, eventually. But I want to move that over once I also move the other interfaces to OCP

File: lib/private/server.php [hunk 2]
```diff
@@ -564,6 +564,26 @@ public function __construct($webRoot, \OC\Config $config) {
...
+				$c->getLogger(),
+				$c->getConfig(),
+				$c->getSecureRandom(),
+				$c->getHasher(),
+				$c->getMountManager(),
+				$c->getGroupManager(),
+				$c->getL10N('core')
+			);
+
+			$manager->addProviders($factory->getProviders());
```
Comment (DeepDiver1975) [inline lib/private/server.php]: hmmm . this would result in creating all providers up front - using the factory within the manager and ask for the explicit provider as needed would have been my approach

Comment (rullzer) [inline lib/private/server.php]: Sure we can do that as well of course. However we need to create all providers at the same time. But that would still be better of course since then it only happens when we actually want to do something with sharing.

Comment (rullzer) [inline lib/private/server.php]: Let me fix this :)

**Comment by rullzer** [comment]
Ok now we call the factory only in the manager. There might be more room for optimization there. But I'd like to move forward with this PR.

CC: @DeepDiver1975 @nickvergessen

File: lib/private/share20/iproviderfactory.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+interface IProviderFactory {
+
+	/**
+	 * Creates and returns the shareProviders
+	 *
+	 * @return IShareProvider[]
+	 */
+	public function getProviders();
+}
```
Comment (LukasReschke) [inline lib/private/share20/iproviderfactory.php]: Newline at EoF

File: lib/private/share20/manager.php [hunk 4]
```diff
@@ -79,13 +88,16 @@ class Manager {
...
 			ILogger $logger,
 			IConfig $config,
-			IShareProvider $defaultProvider,
 			ISecureRandom $secureRandom,
 			IHasher $hasher,
 			IMountManager $mountManager,
 			IGroupManager $groupManager,
-			IL10N $l
+			IL10N $l,
+			IProviderFactory $factory
```
Comment (LukasReschke) [inline lib/private/share20/manager.php]: Missing in the PHPDoc

File: lib/private/share20/manager.php [hunk 5]
```diff
@@ -93,9 +105,79 @@ public function __construct(
...
 		$this->groupManager = $groupManager;
 		$this->l = $l;
+		$this->factory = $factory;
+	}
+
+	/**
+	 * @param IShareProvider[] $providers
+	 * @throws ProviderException
+	 */
+	private function addProviders($providers) {
```
Comment (LukasReschke) [inline lib/private/share20/manager.php]: `array $providers`?

File: lib/private/share20/providerfactory.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	 * @inheritdoc
+	 */
+	public function getProviders() {
+		$providers = [];
+
+		$providers[] = $this->defaultShareProvider();
+
+		return $providers;
+	}
+}
```
Comment (LukasReschke) [inline lib/private/share20/providerfactory.php]: Newline at Eof

File: config/config.sample.php [hunk 7]
```diff
@@ -963,6 +963,22 @@
...
+ *
+ * Global settings for Sharing
+ */
+
+/**
+ * Replaces the default Share Provider Factory. This can be utilized if
+ * own or 3rdParty Share Providers be used that – for instance – uses the
+ * filesystem instead of the database to keep the share information.
+ */
+'sharing.managerFactory' => '\OC\Share20\ProviderFactory',
```
Comment (LukasReschke) [inline config/config.sample.php]: I already see this exploding by all the ~~idiots~~ people that believe that copying our config.sample.php is the way to go, even if it says otherwise :cry:

But fair enough :smile:

File: lib/private/server.php [hunk 8]
```diff
@@ -564,6 +564,25 @@ public function __construct($webRoot, \OC\Config $config) {
 				$request
 			);
 		});
+		$this->registerService('ShareManager', function(Server $c) {
+			$config = $c->getConfig();
+			$factoryClass = $config->getSystemValue('sharing.managerFactory', '\OC\Share20\ProviderFactory');
+			/** @var \OC\Share20\IProviderFactory $factory */
```
Comment (LukasReschke) [inline lib/private/server.php]: :rocket:

File: lib/private/share20/manager.php [hunk 9]
```diff
@@ -584,11 +678,32 @@ public function getShares($page=0, $perPage=50) {
...
+			} catch (ShareNotFound $e) {
+				// Ignore
+			}
+		}
+
+		if ($found === false) {
+			throw new ShareNotFound();
+		}
 
 		return $share;
```
Comment (LukasReschke) [inline lib/private/share20/manager.php]: Makes static code analyzers sad:



While this seems not to be a real world scenario (no provider would need to be defined) it would maybe make sense to check `$found === false || empty($share)`? We can probably also get rid of `$found` completely and just check on `empty($share)`.

**Comment by LukasReschke** [comment]
:+1: besides my nitpicks :smile:

File: config/config.sample.php (hunk 7 above)
Comment (rullzer) [inline config/config.sample.php]: Thinking about it some more. We might remove this from config.sample.php... 99.9999% of the people don't want to change it anyway...

Or maybe just comment it out by default.

Comment (rullzer) [inline config/config.sample.php]: Nah just leave it. Even if people copy the sample things should not break apart.

**Comment by rullzer** [comment]
Comments from @LukasReschke addressed and rebased.

File: lib/private/server.php [hunk 10]
```diff
@@ -1181,5 +1200,12 @@ public function getUserGlobalStoragesService() {
 	public function getUserStoragesService() {
 		return \OC_Mount_Config::$app->getContainer()->query('OCA\\Files_External\\Service\\UserStoragesService');
 	}
-	
+
+
+	/**
+	 * @return \OC\Share20\Manager
```
Comment (nickvergessen) [inline lib/private/server.php]: should be public in the end, so apps can typehint the return

File: lib/private/server.php [hunk 11]
```diff
@@ -1181,5 +1200,12 @@ public function getUserGlobalStoragesService() {
...
+	 */
+	public function getShareManager() {
```
Comment (nickvergessen) [inline lib/private/server.php]: missing from the public interface

File: lib/private/share20/manager.php [hunk 12]
```diff
@@ -93,9 +106,79 @@ public function __construct(
...
+
+	/**
+	 * @param IShareProvider[] $providers
+	 * @throws ProviderException
+	 */
+	private function addProviders(array $providers) {
+		foreach ($providers as $provider) {
+			$name = get_class($provider);
+
+			if (!($provider instanceof IShareProvider)) {
```
Comment (nickvergessen) [inline lib/private/share20/manager.php]: Would have been cool, if share providers could be closures, so they are not constructed when they are not needed, but then other stuff gets more complicated, so either way is fine by me.

File: lib/private/share20/manager.php [hunk 13]
```diff
@@ -93,9 +106,79 @@ public function __construct(
...
+					throw new ProviderException($name . ' tried to register for share type ' . $type . ' but this type is already handled by ' . $this->type2provider[$type]);
+				}
+
+				$this->type2provider[$type] = $name;
+			}
+		}
+	}
+
+	private function runFactory() {
+		if ($this->factoryDone === true) {
```
Comment (nickvergessen) [inline lib/private/share20/manager.php]: could use `!empty($this->providers)` here, and then get rid of the boolean flag?

File: lib/private/server.php (hunk 11 above)
Comment (rullzer) [inline lib/private/server.php]: Yes for now. Until we move everything including the interfaces over..

File: lib/private/share20/manager.php [hunk 14]
```diff
@@ -491,13 +577,21 @@ public function updateShare(Share $share) {
...
+		$this->runFactory();
+
 		$deletedShares = [];
-		foreach($this->defaultProvider->getChildren($share) as $child) {
-			$deletedChildren = $this->deleteChildren($child);
-			$deletedShares = array_merge($deletedShares, $deletedChildren);
 
-			$this->defaultProvider->delete($child);
-			$deletedShares[] = $child;
+		$providerIds = array_keys($this->providers);
```
Comment (nickvergessen) [inline lib/private/share20/manager.php]: why?
```php
foreach($this->providers as $providerId => $provider) {
```

is what you actually want to use below, no need to circle over the keys and then call a method to get the value

Comment (nickvergessen) [inline lib/private/share20/manager.php]: Actually you don't even need the id, so
```php
foreach ($this->providers as $provider) {
```
and your done

File: lib/private/share20/manager.php [hunk 15]
```diff
@@ -584,11 +679,33 @@ public function getShares($page=0, $perPage=50) {
...
 
-		$share = $this->defaultProvider->getShareById($id);
+		//FIXME ids need to become proper providerid:shareid eventually
+
+		$providerIds = array_keys($this->providers);
+		$found = false;
+
+		$share = null;
+		foreach ($providerIds as $providerId) {
+			$provider = $this->getProvider($providerId);
```
Comment (nickvergessen) [inline lib/private/share20/manager.php]: same here again...
```php
foreach ($this->providers as $provider) {
```

File: lib/private/share20/manager.php (hunk 9 above)
Comment (MorrisJobke) [inline lib/private/share20/manager.php]: I just also thought like this. Drop this `$found` ;)

Comment (nickvergessen) [inline lib/private/share20/manager.php]: :+1: for removing found and just defining `$share = false;` before the loop

File: tests/lib/share20/managertest.php [hunk 16]
```diff
@@ -83,22 +89,82 @@ public function setUp() {
...
+				$this->l,
+				$this->factory
+			]);
+	}
+
+	/**
+	 * Initialise the factory. If $providers is null add the defaultProvider
+	 * @param array $providers
+	 */
+	private function facotryProviders(array $providers = null) {
```
Comment (nickvergessen) [inline tests/lib/share20/managertest.php]: faco what? :skull:

File: lib/private/share20/manager.php (hunk 12 above)
Comment (rullzer) [inline lib/private/share20/manager.php]: Yes I agree. We can look at that in a later stage. For now I just want to get this in. But I agree this could be somewhat more elegant.

File: lib/private/share20/manager.php (hunk 13 above)
Comment (rullzer) [inline lib/private/share20/manager.php]: Yes lets do that. Will fix.

File: lib/private/share20/manager.php [hunk 17]
```diff
@@ -588,7 +686,25 @@ public function getShareById($id) {
...
+			try {
+				$share = $provider->getShareById($id);
+				$found = true;
+				break;
+			} catch (ShareNotFound $e) {
+				// Ignore
+			}
+		}
+
+		if ($found === false) {
```
Comment (nickvergessen) [inline lib/private/share20/manager.php]: `$share === null`
and then drop the found

File: config/config.sample.php (hunk 7 above)
Comment (DeepDiver1975) [inline config/config.sample.php]: we shall keep it since documentation is generated based on this file

File: lib/private/share20/manager.php [hunk 18]
```diff
@@ -93,9 +103,88 @@ public function __construct(
...
+		$this->factoryDone = true;
+	}
+
+	/**
+	 * @param string $id
+	 * @return IShareProvider
+	 * @throws ProviderException
+	 */
+	private function getProvider($id) {
+		$this->runFactory();
```
Comment (DeepDiver1975) [inline lib/private/share20/manager.php]: I'd simply forward this call to the factory - no need to build the array.

We need a provider -> ask the factory to build one and we use it.

File: lib/private/share20/manager.php [hunk 19]
```diff
@@ -492,12 +584,17 @@ public function updateShare(Share $share) {
...
 		$deletedShares = [];
-		foreach($this->defaultProvider->getChildren($share) as $child) {
-			$deletedChildren = $this->deleteChildren($child);
-			$deletedShares = array_merge($deletedShares, $deletedChildren);
 
-			$this->defaultProvider->delete($child);
-			$deletedShares[] = $child;
+		$providers = $this->getProviders();
+
+		foreach($providers as $provider) {
```
Comment (DeepDiver1975) [inline lib/private/share20/manager.php]: why do we need to iterate? the share has one share type - we can ask for the provider of the share type and delete the children in there. no need to loop

File: lib/private/share20/manager.php [hunk 20]
```diff
@@ -588,7 +686,24 @@ public function getShareById($id) {
 			throw new ShareNotFound();
 		}
 
-		$share = $this->defaultProvider->getShareById($id);
+		//FIXME ids need to become proper providerid:shareid eventually
+
+		$providers = $this->getProviders();
+
+		$share = null;
+		foreach ($providers as $provider) {
```
Comment (DeepDiver1975) [inline lib/private/share20/manager.php]: same as above - no looping needed

File: lib/private/share20/manager.php (hunk 19 above)
Comment (rullzer) [inline lib/private/share20/manager.php]: Ah yes, left over from the initial approach... but I agree that a share can only have children within its own provider... (actually it will only be used in our provider for user modified group shares).

File: lib/private/share20/manager.php (hunk 20 above)
Comment (rullzer) [inline lib/private/share20/manager.php]: here we need to loop. At least until we add the providerid. Which we can't do in the current situation since that would break OCS... but if I request a share with id 42. That could be 42 for the federated shares or 42 for the 'regular' shares

Comment (rullzer) [inline lib/private/share20/manager.php]: mmm I have an idea how to fix that... so this code is cleaner... let me look into that... but after that I want it merged :stuck_out_tongue_winking_eye:

**Comment by rullzer** [comment]
Warg... this PR got bigger than intented again :(

Now:
* providers are all handled in the factory including fetching logic
* A share now has a fullId (which is 'providerId:shareId'). That way shares are unique. This is required for multiple share providers.
  - However the OCS endpoint v1 can't use them yet since there id's are ints. So we have to hack around it.

One more look please: @DeepDiver1975 @nickvergessen @schiesbn @LukasReschke

File: lib/private/share20/providerfactory.php [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+		throw new ProviderException('No provider with id .' . $id . ' found.');
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function getProviderForType($shareType) {
+		//FIXME we should not report type 2
+		if ($shareType === \OCP\Share::SHARE_TYPE_USER  ||
+			$shareType === 2 ||
```
Comment (rullzer) [inline lib/private/share20/providerfactory.php]: I will fix this in a future PR... doing it here will make it to big.

**Comment by DeepDiver1975** [comment]
move it in :+1: 

any chance in squashing some of these commits?

**Comment by rullzer** [comment]
> any chance in squashing some of these commits?

Done!

File: lib/private/share20/providerfactory.php (hunk 21 above)
Comment (LukasReschke) [inline lib/private/share20/providerfactory.php]: Mind to create an issue for this? - We all know what happens to FIXMEs in the code :wink:

**Comment by LukasReschke** [comment]
:pray: :+1:

**Comment by DeepDiver1975** [comment]
nice
